### PR TITLE
C#: Fix extraction of method signatures

### PR DESCRIPTION
**Pull request by calumgrant**
Fixes a bug whereby method signatures of the form
```
M!0 M!0 FSharp.Core_4.5.0.0::Microsoft.FSharp.Collections.Array4DModule.Get`1(M!0[],Int32,Int32,Int32,Int32)!0[]4[].
Get(Int32,Int32,Int32,Int32);cil-method
```
are decoded incorrectly.

This method signature can be decomposed into
```
T = M!0 FSharp.Core_4.5.0.0::Microsoft.FSharp.Collections.Array4DModule.Get`1(M!0[],Int32,Int32,Int32,Int32)!0

M!0 T[]4[].Get(Int32,Int32,Int32,Int32);cil-method
```
where `T` is a method type-parameter, and `Get(...)` is a method on an array of `T`. What is interesting is that `M!0` on the final line is method type-parameter 0, which is obtained from `T`, and not from the `Get` method which doesn't have type-parameters. Unfortunately, this case was not handled, and the type signature decoder could not find method type-parameter 0 when decoding this signature, and instead threw an exception.

The bug was that the property [`MethodParameters`](https://github.com/calumgrant/ql/blob/526f38c23e2e25e00cb5465634caa51f7d7910a4/csharp/extractor/Semmle.Extraction.CIL/Entities/Method.cs#L30) dropped the method type-parameters in `T` when providing the method type-parameters to the type signature decoder.

The fix is to ensure that the method type-parameters from `T` (contained in the variable `gc`) are correctly inserted into the list of method type-parameters so they are available to the type-signature decoder.

### CODE REVIEW

**Comment by calumgrant** [comment]
This fix has exposed another quite subtle bug in the way that trap IDs are generated. This has caused an inconsistent database that needs to be addressed before this can be merged.

**Comment by pavgust** [comment]
This pull request **introduces 1 alert** when merging b67bc7b61232418673dd9688007c84e7cbfaad17 into 817456ee9fb826bb5aa2d921dd483bdf9b7b302a - [view on LGTM.com](https://lgtm.com/projects/g/Semmle/ql/rev/pr-58d2deb9dee41efdcf14f025e40b653f6877f2f2)

**new alerts:**

* 1 for Reference equality test on System\.Object

---

*Comment posted by [LGTM.com](https://lgtm.com)*

**Comment by calumgrant** [comment]
I've made a few changes, which should be reviewed once the tests go green. I'm not too fussed about whether this makes it into release 1.19. In fact, it may be better to hold off, in which case I'll need to move the change notes into 1.20.

- Fix id generation to qualify method type parameters with their method, to avoid label collision. This fixes the db violations, and gives more constructed generic entities.
- Construct a test case that previously caused the extractor to log an exception, and tests that constructed generics create separate DB entities.

File: csharp/extractor/Semmle.Extraction.CIL/Entities/Type.cs [hunk 1]
```diff
@@ -779,8 +779,8 @@ public ArrayType(Context cx, Type element) : base(cx)
 
         public override Id MakeId(bool inContext) => elementType.GetId(inContext) + openBracket + rank + closeBracket;
 
-        static readonly StringId openBracket = new StringId("[]");
-        static readonly StringId closeBracket = new StringId("[]");
```
Comment (hvitved) [inline csharp/extractor/Semmle.Extraction.CIL/Entities/Type.cs]: Ouch, good catch.

**Pull request merged**